{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "floppy-kuwait",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import csv\n",
    "import os\n",
    "import os.path\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "                           \n",
    "#TSS = int(59)\n",
    "promoter_length = int(100)\n",
    "resultfolder = (\"/home/breena/Documents/Project/blastresult_folder/\")\n",
    "my_data = [f for f in listdir(resultfolder) if isfile(join(resultfolder, f))]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "shaped-kinase",
   "metadata": {},
   "source": [
    "## Extracting the promoter region:\n",
    "Using the blast results to create variables for contigs, strains, gene names and promoter sequence start and end points - because this is a folder of files there is a certain way for reading in folders and then creating a list from the files instead said folder. \n",
    "\n",
    "The 168fna folder read in as handle is a folder of genomes of different strains of Listeria monocytogenes that are available through previous lab work; these files were iterated through to find the names, if the name matched the contig name (defined previously) then the script reads into the master gene array file that I had prepared with all putative SigB-dependent genes present in Listeria monocytogenes, and from this file, any line that contained the gene name of interest was then iterated through to pull out the distance from the transcription start site (TSS) which was previously recorded using Geneious. \n",
    "\n",
    "A folder of genbank formatted files of these same 168 strains was then used to iterate through in order to find the record id that matched previously defined contig names and then if a hit was found the promoter_start and promoter_end points were recorded for both positive and complementary strands; these positions were used then to iterate through the files and gather the promoter sequences that were 50 bp in length. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "expanded-concern",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lmo0211\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "import csv\n",
    "import os.path\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "path = \"/home/breena/Documents/Project/blastresult_folder/\"\n",
    "data = [f for f in listdir(path) if isfile(join(path, f))]\n",
    "for file in data:\n",
    "    print(file)\n",
    "#path_1 = (\"/home/breena/Documents/Project/blastresult_folder/\")\n",
    "#files = os.listdir(path_1)\n",
    "#for f in files:\n",
    "    #print(f)\n",
    "    promoter_length = int(50) #change this so that the promoter seq is only 50 bp instead of 200\n",
    "    #with open(path + file, \"r\") as resultfile:\n",
    "            file_path = (path + file).replace(\".fasta_output.csv\", \"\")\n",
    "            #print(path)\n",
    "            gene_name = os.path.basename(file_path)\n",
    "            print(gene_name)\n",
    "            output = open(\"SigB_P\" + gene_name + \".fasta\", \"w+\")\n",
    "            for line in resultfile:\n",
    "                contig = line.split(\"\\t\")[4]\n",
    "                #print(contig)\n",
    "                contig_name = contig.split(\"|\")[2]\n",
    "                #print(contig_name)\n",
    "                strain_name = contig_name.split(\"_\")[0]\n",
    "                #print(strain_name)\n",
    "                start = int(line.split(\"\\t\")[2])\n",
    "                #print(start)\n",
    "                end = int(line.split(\"\\t\")[3])\n",
    "                #print(end)\n",
    "                if (int(end) - int(start)) >=0:\n",
    "                    hit_strand = \"+\"\n",
    "                else:\n",
    "                    hit_strand = \"-\"\n",
    "                with open(\"/home/breena/Documents/Project/168fna/\" + strain_name + \".fna\") as handle:\n",
    "                    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "                    #print(record.id)\n",
    "                    #print(repr(record.seq))\n",
    "                    #print(len(record))\n",
    "                        name = (record.id).split(\"|\")[2]\n",
    "                        #print(name)\n",
    "                        if name == contig_name:\n",
    "                            #print(\"contig found \" + contig_name + hit_strand)\n",
    "                            with open(\"/home/breena/Documents/Project/genes.csv\", \"r\") as infile:\n",
    "                                data_reader = csv.reader(infile)\n",
    "                                for line in infile:\n",
    "                                    #print(gene_name)\n",
    "                                    if gene_name in line:\n",
    "                                    #if line.startswith(gene_name):\n",
    "                                        distance = int(line.split(\",\")[3])\n",
    "                                        #print(distance)\n",
    "                                        #break\n",
    "                                        #using genbank files from here - should be fasta?\n",
    "                                        gbkfolder = \"/home/breena/Documents/Project/168gbk/\"\n",
    "                                        #my_data = [f for f in listdir(gbkfolder) if isfile(join(gbkfolder, f))]\n",
    "                                        #the distance, gene_name should be taken from each row in the database file (columns A and E)\n",
    "                                        #for file in my_data:\n",
    "                                            #print(file)\n",
    "                                        for rec in SeqIO.parse(gbkfolder + strain_name + \".gbk\", \"genbank\"):\n",
    "                                            #print(rec.id)\n",
    "                                            if rec.id == contig_name:\n",
    "                                                if hit_strand == \"+\":\n",
    "                                                    promoter_start = int(start - distance - promoter_length)\n",
    "                                                    promoter_end = int(start - distance)\n",
    "                                                    #promoter = SeqFeature(FeatureLocation(280913,282107), type = \"SigB-promotor\", strand = 1)\n",
    "                                                    promoter = SeqFeature(FeatureLocation(promoter_start, promoter_end),\n",
    "                                                                          type=\"SigB-\" + gene_name + \"-promotor\", strand = 1)\n",
    "                                                    #print(promoter)\n",
    "                                                    promoter_seq = (promoter.location.extract(rec).seq)\n",
    "                                                    #print(promoter_seq)                                             \n",
    "                                                else:\n",
    "                                                    promoter_start = int(start + distance + promoter_length)\n",
    "                                                    promoter_end = int(start + distance)\n",
    "                                                    promoter = SeqFeature(FeatureLocation(promoter_end, promoter_start),\n",
    "                                                                          type=\"SigB-\" + gene_name + \"-promotor\", strand = -1)\n",
    "                                                    #print(promoter)\n",
    "                                                    promoter_seq = promoter.location.extract(rec).seq\n",
    "                                                    #print(promoter_seq)\n",
    "                                                    #promotor = SeqFeature(FeatureLocation(promoter_start, promoter_end), strand=\"+\")\n",
    "                                                        #promotors.append(promotor.location.extract(rec).seq\n",
    "                                                #option 1 for writing output to file\n",
    "\n",
    "                                                output.write(str('>' + strain_name.replace(\".fna\",\"\") + '_SigB_P' + gene_name + '\\n' + promoter_seq + '\\n'))\n",
    "\n",
    "            output.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "starting-subsection",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
